### PR TITLE
AVRO-2357: Allow generic types in reflect protos

### PR DESCRIPTION
**Pull request by ivangreene**
Adds support for generic types in ReflectData for Protocols.

I also took the liberty of cleaning up the iteration over Parameters; the multi-dimensional array style is no longer really necessary with the new methods on Parameter.

https://issues.apache.org/jira/browse/AVRO-2357

### CODE REVIEW

**Comment by iemejia** [comment]
We introduced automatic code formatting. For more info see the "How to contribute" page. This  probably affected this PR can you please rebase it.

**Comment by ivangreene** [comment]
@iemejia thanks for the heads up. I just rebased and applied spotless formatting.

File: lang/java/avro/src/main/java/org/apache/avro/reflect/ReflectData.java [hunk 1]
```diff
@@ -769,12 +770,13 @@ public Protocol getProtocol(Class iface) {
         iface.getPackage() == null ? "" : iface.getPackage().getName());
     Map<String, Schema> names = new LinkedHashMap<>();
     Map<String, Message> messages = protocol.getMessages();
+    Map<TypeVariable<?>, Type> genericTypeVariableMap = ReflectionUtil.resolveTypeVariables(iface);
     for (Method method : iface.getMethods())
```
Comment (Fokko) [inline lang/java/avro/src/main/java/org/apache/avro/reflect/ReflectData.java]: Not part of the PR, but can we add brackets here, looks awkward.

Comment (ivangreene) [reply lang/java/avro/src/main/java/org/apache/avro/reflect/ReflectData.java]: Done

Comment (Fokko) [reply lang/java/avro/src/main/java/org/apache/avro/reflect/ReflectData.java]: Thanks!
